File: permissionx/src/main/java/com/lws/permissionx/PermissionResult.java
package com.lws.permissionx;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

/**
 * @author lws
 */
public class PermissionResult {

    PermissionResult() {
    }

    void addGranted(String permission) {
        grantedList.add(permission);
    }

    void addDenied(String permission) {
        deniedList.add(permission);
    }

    private final List<String> grantedList = new ArrayList<>();
    private final List<String> deniedList = new ArrayList<>();

    public boolean isAllGranted() {
        return deniedList.isEmpty();
    }

    public List<String> getGrantedList() {
        return grantedList;
    }

    public List<String> getDeniedList() {
        return deniedList;
    }

    @NonNull
    @Override
    public String toString() {
        return "PermissionResult{" +
                "grantedList=" + grantedList +
                ", deniedList=" + deniedList +
                '}';
    }
}

File: permissionx/src/main/java/com/lws/permissionx/PermissionResult.kt
package com.lws.permissionx

/**
 * @author lws
 */
class PermissionResult internal constructor(result: Map<String, Boolean>) {
    private val grantedList: MutableList<String> = ArrayList()
    private val deniedList: MutableList<String> = ArrayList()

    init {
        for (permission in result.keys) {
            if (java.lang.Boolean.TRUE == result[permission]) {
                grantedList.add(permission)
            } else {
                deniedList.add(permission)
            }
        }
    }

    val isAllGranted: Boolean
        get() = deniedList.isEmpty() && grantedList.isNotEmpty()

    fun getGrantedList(): List<String> {
        return grantedList
    }

    fun getDeniedList(): List<String> {
        return deniedList
    }

    override fun toString(): String {
        return "PermissionResult{" +
                "grantedList=" + grantedList +
                ", deniedList=" + deniedList +
                '}'
    }
}

fun interface PermissionResultCallback {
    fun onPermissionResult(result: PermissionResult)
}